api/v1alpha1/addresspool_webhook.go

```diff
@@ -0,0 +1,168 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/mikioh/ipaddr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging addresspool-webhook.
+var addresspoollog = logf.Log.WithName("addresspool-webhook")
+var c client.Client
+
+func (addressPool *AddressPool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	c = mgr.GetClient()
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(addressPool).
+		Complete()
+}
+
+//+kubebuilder:webhook:verbs=create;update;delete,path=/validate-metallb-io-v1alpha1-addresspool,mutating=false,failurePolicy=fail,groups=metallb.io,resources=addresspools,versions=v1alpha1,name=addresspoolvalidationwebhook.metallb.io
+var _ webhook.Validator = &AddressPool{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateCreate() error {
+	addresspoollog.Info("validate AddressPool creation", "name", addressPool.Name)
+
+	return addressPool.validateAddressPool(true)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateUpdate(old runtime.Object) error {
+	addresspoollog.Info("validate AddressPool update", "name", addressPool.Name)
+
+	return addressPool.validateAddressPool(false)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateDelete() error {
+	addresspoollog.Info("validate AddressPool deletion", "name", addressPool.Name)
+
+	return nil
+}
+
+func cidrsOverlap(a, b *net.IPNet) bool {
+	return cidrContainsCIDR(a, b) || cidrContainsCIDR(b, a)
+}
+
+func cidrContainsCIDR(outer, inner *net.IPNet) bool {
+	ol, _ := outer.Mask.Size()
+	il, _ := inner.Mask.Size()
+	if ol == il && outer.IP.Equal(inner.IP) {
+		return true
+	}
+	if ol < il && outer.Contains(inner.IP) {
+		return true
+	}
+	return false
+}
+
+func parseCIDR(cidr string) ([]*net.IPNet, error) {
+	if !strings.Contains(cidr, "-") {
+		_, n, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid CIDR %q", cidr)
+		}
+		return []*net.IPNet{n}, nil
+	}
+
+	fs := strings.SplitN(cidr, "-", 2)
+	if len(fs) != 2 {
+		return nil, fmt.Errorf("invalid IP range %q", cidr)
+	}
+	start := net.ParseIP(strings.TrimSpace(fs[0]))
+	if start == nil {
+		return nil, fmt.Errorf("invalid IP range %q: invalid start IP %q", cidr, fs[0])
+	}
+	end := net.ParseIP(strings.TrimSpace(fs[1]))
+	if end == nil {
+		return nil, fmt.Errorf("invalid IP range %q: invalid end IP %q", cidr, fs[1])
+	}
+
+	var ret []*net.IPNet
+	for _, pfx := range ipaddr.Summarize(start, end) {
```

Comment:
Couple of things here:
- it's not entirely clear to me how this is supposed to work. My understanding is that this is returning all the combination of cidrs of a given range
- We should be careful in bringing in new dependencies (unless they are well known k8s.io/util or from the std library)

I see a couple of alternatives:
- differentiate from when we parse a cidr and when we parse a range and implement comparison functions for both (cidr vs cidr, range vs range, cidr vs range). The benefit here is that we have more control and I believe the code will be more readable
- add more comments to explain what is happening here (plus the aforementioned unit tests)

Even if it will take more, I am inclined to favor the first approach.

Reply:
this code is taken from metallb config, if we ever move the crds and the webhook to metallb we would like to use the same code. so for now let's keep it as it is.

Reply:
Sounds good to me. Let's put all the stolen code under the same file with a big TODO "refactor when we move the crds and the webhook to metallb"